### PR TITLE
Create PyPI Package and Separate Training/Client Code

### PR DESCRIPTION
**Pull request by edwinzhng**
1. Refactored original howl code into `howl` and `howl-training` folders (moved all files under `align` and `run` into `howl-training` and separated the dependencies into two files)
2. Created a basic HowlClient for the PyPI package to load pretrained models from PyTorch hub. This currently only supports "hey_fire_fox" as a pretrained model name, but users can provide their own manually as well

Example usage (also in README):
```
from howl.client import HowlClient

def hello_callback(detected_words):
    print("Detected: {}".format(detected_words))

client = HowlClient()
client.from_pretrained("hey_fire_fox")
client.add_listener(hello_callback)
client.start().join()
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -18,61 +20,78 @@ Citation:
...
+### Installation
 
 1. `git clone https://github.com/castorini/howl && cd howl`
 
 2. Install [PyTorch](https://pytorch.org) by following your platform-specific instructions.
 
 3. Install PyAudio and its dependencies through your distribution's package system.
 
-4. `pip install -r requirements.txt` (some apt packages might need to be installed)
+4. `pip install -r requirements.txt -r requirements_training.txt` (some apt packages might need to be installed)
```
Comment (ljj7975) [inline README.md:39]: does this work `pip install -r requirements.txt requirements_training.txt`?

File: howl/client/howl_client.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+import logging
+import time
+from typing import Callable
+
+import numpy as np
+import pyaudio
+import torch
+from howl.model.inference import InferenceEngine
+
+
```
Comment (ljj7975) [inline howl/client/howl_client.py:11]: can you add some comments about what this class is for?

File: howl/client/howl_client.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+class HowlClient:
+    def __init__(self,
+                 engine: InferenceEngine = None,
+                 context: InferenceContext = None,
+                 device: int = -1,
```
Comment (ljj7975) [inline howl/client/howl_client.py:16]: what is device for?

File: howl/client/howl_client.py (hunk 2 above)
Comment (ljj7975) [reply howl/client/howl_client.py:11]: maybe we should adopt the idea of docstring

File: howl/client/howl_client.py (hunk 3 above)
Comment (ljj7975) [reply howl/client/howl_client.py:16]: can't this just be a string?

File: howl/client/howl_client.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    def _on_audio(self, in_data):
+        data_ok = (in_data, pyaudio.paContinue)
+        self.last_data = in_data
+        self._audio_buf.append(in_data)
+        if len(self._audio_buf) != 16:
+            return data_ok
+
+        audio_data = b''.join(self._audio_buf)
+        self._audio_buf = self._audio_buf[2:]
+        arr = np.frombuffer(audio_data, dtype=np.int16).astype(np.float) / 32767
```
Comment (ljj7975) [inline howl/client/howl_client.py:48]: can we make 32767 a constant somewhere?

Comment (ljj7975) [reply howl/client/howl_client.py:48]: maybe class variable

File: howl/client/howl_client.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        if device == -1:
+            return torch.device('cpu')
+        else:
+            return torch.device('cuda:{}'.format(device))
+
+    def _on_audio(self, in_data):
+        data_ok = (in_data, pyaudio.paContinue)
+        self.last_data = in_data
+        self._audio_buf.append(in_data)
+        if len(self._audio_buf) != 16:
```
Comment (ljj7975) [inline howl/client/howl_client.py:43]: const as well

File: setup.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+with open('requirements.txt') as f:
+    requirements = f.read().splitlines()
+
+setuptools.setup(
+    name="howl",
+    version="0.0.1",
+    author="Anserini Gaggle",
+    author_email="anserini.gaggle@gmail.com",
+    description="A wake word detection modeling toolkit",
```
Comment (ljj7975) [inline setup.py:14]: I would say `A wake word detection toolkit`

File: hubconf.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+import torch
+
+from howl.context import InferenceContext
+from howl.data.transform import ZmuvTransform
+from howl.model import RegisteredModel, Workspace
+from howl.model.inference import (FrameInferenceEngine, InferenceEngine,
+                                  SequenceInferenceEngine)
+from howl.settings import SETTINGS
+
+
```
Comment (ljj7975) [inline hubconf.py:15]: can you add some comments about what this file is for and how it is used?

File: README.md (hunk 1 above)
Comment (edwinzhng) [reply README.md:39]: Yeah you can specify multiple

File: howl/client/howl_client.py (hunk 2 above)
Comment (edwinzhng) [reply howl/client/howl_client.py:11]: yeah still adding some code, will add docs after

File: howl/client/howl_client.py (hunk 4 above)
Comment (edwinzhng) [reply howl/client/howl_client.py:48]: Sure, any naming suggestions? Copied this from demo.py

File: howl/client/howl_client.py (hunk 5 above)
Comment (edwinzhng) [reply howl/client/howl_client.py:43]: Any naming suggestion for this too? Also just took it from demo.py, should I maybe pull both out and reuse it in both files?

File: howl/client/howl_client.py (hunk 3 above)
Comment (edwinzhng) [reply howl/client/howl_client.py:16]: Added a comment "Device for CPU/GPU support. A value of -1 defaults to CPU, positives values will run the model on the associated CUDA device ID.". Would it be fine to ignore multi GPU support for now? Eventually we could change it to pass an array of multiple devices

File: howl/client/howl_client.py (hunk 4 above)
Comment (edwinzhng) [reply howl/client/howl_client.py:48]: Actually I just defined this for now
```
self._audio_buf_len = 16
self._audio_float_size = 32767
```

File: howl/client/howl_client.py (hunk 2 above)
Comment (edwinzhng) [reply howl/client/howl_client.py:11]: ok added some docs, let me know if anything is unclear to you

File: README.md [hunk 8]
```diff
@@ -18,61 +20,93 @@ Citation:
...
+2. Install Howl using `pip`
+
+```
+pip install howl
+```
+
+3. To immediately use a pre-trained Howl model for inference, we provide the `client` API. The following example loads the "hey_fire_fox" pretrained model with a simple callback and starts the inference client.
+
+```
+from howl.client import HowlClient
```
Comment (edwinzhng) [inline README.md:34]: Here is the example for testing

File: README.md (hunk 1 above)
Comment (ljj7975) [reply README.md:55]: You have two `-r` so I was asking

File: howl/client/howl_client.py (hunk 3 above)
Comment (ljj7975) [reply howl/client/howl_client.py:24]: yea we can ignore multi gpu for now

File: README.md (hunk 8 above)
Comment (ljj7975) [reply README.md:34]: I think we should have a file for this along with this code in the Readme

Comment (edwinzhng) [reply README.md:34]: Oh you mean like a basic demo file?

Comment (edwinzhng) [reply README.md:34]: Added a file to examples/hey_fire_fox.py

Comment (ljj7975) [reply README.md:34]: can you update the read me?

File: examples/hey_fire_fox.py [hunk 9]
```diff
@@ -0,0 +1,9 @@
+from howl.client import HowlClient
+
```
Comment (ljj7975) [inline examples/hey_fire_fox.py:2]: Can you write some comments about what this file is supposed to do.?
just like how I had it for last scripts that got merged?

File: howl/__init__.py [hunk 10]
```diff
@@ -1 +1,2 @@
-from .logger import *
+from howl.client import *
```
Comment (ljj7975) [inline howl/__init__.py:1]: we need this here?

File: howl/__init__.py [hunk 11]
```diff
@@ -1 +1,2 @@
...
+from howl.logger import *
```
Comment (ljj7975) [inline howl/__init__.py:2]: I feel like this file is unnecessary. can you double check?

File: howl/client/howl_client.py [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+        """Block while the audio inference stream is active"""
+        while self.stream.is_active():
+            time.sleep(0.1)
+
+    def from_pretrained(self, name: str, force_reload: bool = False):
+        """Load a pretrained model using the provided name"""
+        engine, ctx = torch.hub.load(
+            'castorini/howl:howl-pip', name, force_reload=force_reload, reload_models=force_reload)
+        self.engine = engine.to(self.device)
+        self.ctx = ctx
```
Comment (ljj7975) [inline howl/client/howl_client.py:130]: I know we had two parameters incorrectly setup previously.
`INFERENCE_THRESHOLD=0 NUM_MELS=40`
we should be loading these directly from these pre-trained models

File: howl/model/inference.py [hunk 13]
```diff
@@ -125,10 +126,11 @@ def __init__(self,
                  negative_label: int,
                  settings: InferenceEngineSettings = InferenceEngineSettings(),
                  coloring: LabelColoring = None,
-                 time_provider=time.time):
+                 time_provider=time.time,
+                 audio_transform_settings: AudioTransformSettings = AudioTransformSettings()):
```
Comment (ljj7975) [inline howl/model/inference.py:130]: I believe this is what enables getting the right parameters. Am I correct?

File: hubconf.py [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+    """
+    Pretrained model for Firefox Voice
+    """
+    # Audio inference settings
+    conf = _ModelSettings(path="howl/hey-fire-fox",
+                          model_name="res8",
+                          vocab=["hey", "fire", "fox"],
+                          inference_sequence=[0, 1, 2],
+                          token_type="word",
+                          use_frame=True)
```
Comment (ljj7975) [inline hubconf.py:48]: I think we need to design this one little more carefully.

I think what we eventually want is other people developing training their own model, uploading it to hub, and then they should update this code in a minimal way to expose their models through the pip package.

Having said that, I feel like we need a helper function that does model loading and hey_fire_fox should only be creating the setting and triggering that helper function.

So when someone else add new model, all they need to do is simply creating a new setting and triggering that helper

File: hubconf.py [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+    ats = transform.augment.AudioTransformSettings(num_mels=40)
+    _SETTINGS._training = settings.TrainingSettings(max_window_size_seconds=0.5)
```
Comment (ljj7975) [inline hubconf.py:50]: these numbers must be loaded from the load_model function. (i.e. should never be hard coded)
a stored model has dict and as long as we are storing those information, you should see it when you load the model.

File: hubconf.py [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+
+    # Separate `reload_models` flag since PyTorch will pop the 'force_reload' flag
+    reload_models = kwargs.pop('reload_models', False)
+    _download_howl_models(cache_dir, reload_models)
+
+    # Set up context and workspace
+    ws_path: pathlib.Path = cache_dir / _MODEL_CACHE_FOLDER / conf.path
+    ctx = context.InferenceContext(
+        conf.vocab, token_type=conf.token_type, use_blank=not conf.use_frame)
+    ws = howl_model.Workspace(ws_path, delete_existing=False)
```
Comment (ljj7975) [inline hubconf.py:65]: check this object for getting the right parameters (num_mels, inference sequence etc)

File: howl/model/workspace.py [hunk 17]
```diff
@@ -44,4 +54,15 @@ def save_model(self, model: nn.Module, best=False):
...
     def load_model(self, model: nn.Module, best=True):
-        model.load_state_dict(torch.load(self.model_path(best=best), lambda s, l: s))
+        model.load_state_dict(torch.load(
+            self.model_path(best=best), lambda s, l: s))
+
+    def write_settings(self, settings):
+        with (self.path / 'settings.json').open('w') as f:
+            json.dump(gather_dict(settings), f, indent=2)
+
+    def load_settings(self) -> WorkspaceSettings:
```
Comment (edwinzhng) [inline howl/model/workspace.py:64]: I can change this to read a `.env` file

**Pull request merged**